### PR TITLE
plugin/grpc: New gRPC plugin

### PR DESCRIPTION
**Pull request by inigohu**
### 1. Why is this pull request needed and what does it do?
For proxying DNS messages to upstream resolvers using gRPC connection

### 2. Which issues (if any) are related?
#1476 
### 3. Which documentation changes (if any) need to be made?
-
### 4. Does this introduce a backward incompatible change or deprecation?
No

### CODE REVIEW

File: plugin/grpc/grpc.go
Comment (miekg) [inline plugin/grpc/grpc.go:12]: seperate coredns imports from 3rd party with a newline

File: plugin/grpc/grpc.go
Comment (miekg) [inline plugin/grpc/grpc.go]: comment needs adjusting

File: plugin/grpc/proxy.go
Comment (miekg) [inline plugin/grpc/proxy.go]: should probably just return the error instead of logging or continuing?

File: plugin/grpc/proxy.go
Comment (miekg) [inline plugin/grpc/proxy.go]: doesn't gRPC have all kinds of stuff for this? I.e reconnecting when the connection breaks etc?

File: plugin/grpc/proxy_test.go
Comment (miekg) [inline plugin/grpc/proxy_test.go]: would just call it testServiceClient or some such.

File: plugin/grpc/setup.go
Comment (miekg) [inline plugin/grpc/setup.go]: doesn't have to be public

**Review by miekg — COMMENTED** [review]
thanks!

there is no of course some code duplication between forward and grpc - which is fine, but it does make sense to cut back on that a little (there will at some point be a 'doh' proxy as well.

I'm fine with delaying de-dupping this, after this is merged.

File: plugin/grpc/proxy.go
Comment (inigohu) [reply plugin/grpc/proxy.go]: You're right. I've deleted all **healtcheck** stuff.

**Comment by inigohu** [comment]
I've this error in the CI 

```
build github.com/coredns/coredns: cannot find module for path github.com/DataDog/dd-trace-go/tracer
Makefile:17: recipe for target 'coredns' failed
make[1]: *** [coredns] Error 1
```

I've realized that in _go.mod_ (master branch) file are 3 dependencies referencing to _DataDog/dd-trace-go_ . This may be the cause

```
github.com/DataDog/dd-trace-go v0.6.1
gopkg.in/DataDog/dd-trace-go.v0 v0.6.1
gopkg.in/DataDog/dd-trace-go.v1 v1.10.0
```

**Comment by miekg** [comment]
Latest version fixes that. Fighting with it all morning

**Comment by miekg** [comment]
rebasing on top of master (think you tried that w/ the force push) should fix it

**Comment by inigohu** [comment]
> rebasing on top of master (think you tried that w/ the force push) should fix it

The branch is up to date with master and still get an error.

**Comment by miekg** [comment]
[ Quoting <notifications@github.com> in "Re: [coredns/coredns] plugin/grpc: ..." ]
>I've this error in the CI
>
>```
>build github.com/coredns/coredns: cannot find module for path github.com/DataDog/dd-trace-go/tracer
>Makefile:17: recipe for target 'coredns' failed
>make[1]: *** [coredns] Error 1
>```
>
>I've realized that in _go.mod_ (master branch) file are 3 dependencies referencing to _DataDog/dd-trace-go_ . This may be the cause
>
>```
>github.com/DataDog/dd-trace-go v0.6.1
>gopkg.in/DataDog/dd-trace-go.v0 v0.6.1
>gopkg.in/DataDog/dd-trace-go.v1 v1.10.0
>```

go modules has a learning curve.

anyway, things in master should be better now.

**Comment by lopezator** [comment]
np @miekg we're working together. thanks!

File: plugin.cfg
Comment (miekg) [inline plugin.cfg]: this should be near *forward*, to keep forwarders together. Order matters here.

**Comment by miekg** [comment]
this also needs an owner file with you in it, which also implies I will add you as a maintainer to the coredns project. (Unless you object, which is fine, but then we need to discuss what we do here)

**Comment by inigohu** [comment]
> this also needs an owner file with you in it, which also implies I will add you as a maintainer to the coredns project. (Unless you object, which is fine, but then we need to discuss what we do here)

Of course, you can add me as a maintainer.
Thanks

**Comment by miekg** [comment]
[ Quoting <notifications@github.com> in "Re: [coredns/coredns] plugin/grpc: ..." ]
>> this also needs an owner file with you in it, which also implies I will add you as a maintainer to the coredns project. (Unless you object, which is fine, but then we need to discuss what we do here)
>
>Of course, you can add me as a maintainer.
>Thanks

invite send

**Comment by inigohu** [comment]
> invite send

@miekg do you want to make a review or can I merge it?

**Comment by miekg** [comment]
I wanna do a last pass

File: plugin/grpc/grpc.go
Comment (miekg) [inline plugin/grpc/grpc.go]: really, we do this this in forward as well? (assuming the code is copied). This odd

**Review by miekg — COMMENTED** [review]
one question, for the rest lgtm

File: plugin/grpc/grpc.go
Comment (inigohu) [reply plugin/grpc/grpc.go]: No, I added this check to continue searching in other proxies if no results has been found.

**Comment by miekg** [comment]
[ Quoting <notifications@github.com> in "Re: [coredns/coredns] plugin/grpc: ..." ]
>+		// Continue if no answer has been found
>+		if ret.Rcode == dns.RcodeNameError {
>
>No, I added this check to continue searching in other proxies if no results has been found.

that's not the job of the proxy and can be removed

**Review by miekg — APPROVED** [review]

**Pull request merged**
